File: esp32/src/watchdog.cpp
#include "watchdog.h"
#include "wolf_cwl.h"
#include "logging.h"
#include <esp_task_wdt.h>
#include <esp_system.h>
#include <Preferences.h>

// =============================================================================
// Configuration
// =============================================================================
#define WDT_TIMEOUT_S        30       // Hardware watchdog: 30 seconds
#define OT_LIVENESS_TIMEOUT  300000   // OpenTherm: restart after 5 minutes no response
#define OT_STARTUP_GRACE     60000    // Wait 60s after boot before checking OT liveness
#define HEAP_MIN_BYTES       20480    // Restart if free heap drops below 20 KB
#define LIVENESS_CHECK_INTERVAL 10000 // Run soft checks every 10 seconds

// =============================================================================
// State
// =============================================================================
static uint32_t crashCount = 0;
static esp_reset_reason_t resetReason;
static unsigned long lastLivenessCheck = 0;
static unsigned long bootTime = 0;

// =============================================================================
// Reboot reason decoder
// =============================================================================
const char* getRebootReasonStr() {
    switch (resetReason) {
        case ESP_RST_POWERON:  return "Power";
        case ESP_RST_EXT:      return "External";
        case ESP_RST_SW:       return "SW";
        case ESP_RST_PANIC:    return "Panic";
        case ESP_RST_INT_WDT:  return "IntWDT";
        case ESP_RST_TASK_WDT: return "TaskWDT";
        case ESP_RST_WDT:      return "WDT";
        case ESP_RST_DEEPSLEEP:return "DeepSleep";
        case ESP_RST_BROWNOUT: return "Brownout";
        case ESP_RST_SDIO:     return "SDIO";
        default:               return "Unknown";
    }
}

static bool isAbnormalReset(esp_reset_reason_t reason) {
    return reason == ESP_RST_PANIC ||
           reason == ESP_RST_INT_WDT ||
           reason == ESP_RST_TASK_WDT ||
           reason == ESP_RST_WDT ||
           reason == ESP_RST_BROWNOUT;
}

// =============================================================================
// Setup
// =============================================================================
void setupWatchdog() {
    bootTime = millis();
    resetReason = esp_reset_reason();

    // Load crash counter from NVS
    Preferences prefs;
    prefs.begin("watchdog", true);
    crashCount = prefs.getUInt("crash_count", 0);
    prefs.end();

    // Increment crash counter on abnormal reset
    if (isAbnormalReset(resetReason)) {
        crashCount++;
        Preferences prefsW;
        prefsW.begin("watchdog", false);
        prefsW.putUInt("crash_count", crashCount);
        prefsW.end();
        log("Watchdog: Abnormal reset detected (" + String(getRebootReasonStr()) +
            "), crash count: " + String(crashCount));
    } else {
        log("Watchdog: Boot reason: " + String(getRebootReasonStr()) +
            ", crash count: " + String(crashCount));
    }

    // Initialize hardware Task Watchdog Timer
    esp_task_wdt_init(WDT_TIMEOUT_S, true);  // true = trigger reset on timeout
    esp_task_wdt_add(NULL);  // Add current task (loopTask)

    log("Watchdog: TWDT enabled (" + String(WDT_TIMEOUT_S) + "s timeout)");
}

// =============================================================================
// Feed watchdog + software liveness checks
// =============================================================================
void feedWatchdog() {
    // Feed hardware watchdog
    esp_task_wdt_reset();

    // Run soft liveness checks periodically
    unsigned long now = millis();
    if (now - lastLivenessCheck < LIVENESS_CHECK_INTERVAL) return;
    lastLivenessCheck = now;

    // OpenTherm liveness check (with startup grace period)
    if (now - bootTime > OT_STARTUP_GRACE) {
        if (cwlData.lastResponse > 0) {
            unsigned long otAge = now - cwlData.lastResponse;
            if (otAge > OT_LIVENESS_TIMEOUT) {
                log("Watchdog: OpenTherm unresponsive for " + String(otAge / 1000) +
                    "s, restarting...");
                delay(100);
                ESP.restart();
            }
        } else {
            // Never received a response — check if we've been waiting too long
            unsigned long waitTime = now - bootTime;
            if (waitTime > OT_LIVENESS_TIMEOUT) {
                log("Watchdog: No OpenTherm response since boot (" +
                    String(waitTime / 1000) + "s), restarting...");
                delay(100);
                ESP.restart();
            }
        }
    }

    // Heap exhaustion check
    uint32_t freeHeap = ESP.getFreeHeap();
    if (freeHeap < HEAP_MIN_BYTES) {
        log("Watchdog: Heap critically low (" + String(freeHeap) +
            " bytes), restarting...");
        delay(100);
        ESP.restart();
    }
}

// =============================================================================
// Diagnostics getters
// =============================================================================
uint32_t getCrashCount() {
    return crashCount;
}

unsigned long getUptimeSeconds() {
    return millis() / 1000;
}

unsigned long getOtResponseAge() {
    if (cwlData.lastResponse == 0) return millis() / 1000;  // Never received
    return (millis() - cwlData.lastResponse) / 1000;
}

File: esp32/src/watchdog.rs
//! Watchdog — task WDT, liveness checks, crash counter.

use esp_idf_svc::sys::*;
use log::{info, warn};

const WDT_TIMEOUT_S: u32 = 30;
const LIVENESS_INTERVAL_MS: u32 = 10_000;
const OT_TIMEOUT_MS: u32 = 300_000; // 5 minutes
const MIN_FREE_HEAP: u32 = 20_480;  // 20 KB

pub struct Watchdog {
    last_check_ms: u32,
    startup_grace_ms: u32,
}

impl Watchdog {
    pub fn new() -> Self {
        unsafe {
            let config = esp_task_wdt_config_t {
                timeout_ms: WDT_TIMEOUT_S * 1000,
                idle_core_mask: 0,
                trigger_panic: true,
            };
            // ESP-IDF already initializes TWDT — just reconfigure with our timeout
            esp_task_wdt_reconfigure(&config);
            // Add current task to WDT
            let handle = xTaskGetCurrentTaskHandle();
            let ret = esp_task_wdt_add(handle);
            if ret != ESP_OK {
                // Task might already be added — delete and re-add
                esp_task_wdt_delete(handle);
                esp_task_wdt_add(handle);
            }
        }
        info!("Watchdog: TWDT enabled ({}s timeout)", WDT_TIMEOUT_S);

        Self {
            last_check_ms: 0,
            startup_grace_ms: 60_000,
        }
    }

    /// Feed the watchdog and run liveness checks. Call from main loop.
    pub fn update(&mut self, now_ms: u32, ot_last_response_ms: u32, ot_connected: bool) {
        // Feed the watchdog
        unsafe { esp_task_wdt_reset() };

        if now_ms < self.startup_grace_ms {
            return; // Grace period after boot
        }

        if now_ms.wrapping_sub(self.last_check_ms) < LIVENESS_INTERVAL_MS {
            return;
        }
        self.last_check_ms = now_ms;

        // Check OpenTherm liveness
        if ot_connected && now_ms.wrapping_sub(ot_last_response_ms) > OT_TIMEOUT_MS {
            warn!("Watchdog: OpenTherm not responding for 5 min, restarting");
            unsafe { esp_restart() };
        }

        // Check heap
        let free_heap = unsafe { esp_get_free_heap_size() };
        if free_heap < MIN_FREE_HEAP {
            warn!("Watchdog: Low heap ({}), restarting", free_heap);
            unsafe { esp_restart() };
        }
    }
}

/// Get reboot reason as string
pub fn reboot_reason() -> &'static str {
    #[allow(non_upper_case_globals)]
    match unsafe { esp_reset_reason() } {
        esp_reset_reason_t_ESP_RST_POWERON => "Power",
        esp_reset_reason_t_ESP_RST_SW => "SW",
        esp_reset_reason_t_ESP_RST_PANIC => "Panic",
        esp_reset_reason_t_ESP_RST_INT_WDT => "IntWDT",
        esp_reset_reason_t_ESP_RST_TASK_WDT => "TaskWDT",
        esp_reset_reason_t_ESP_RST_WDT => "WDT",
        esp_reset_reason_t_ESP_RST_BROWNOUT => "Brownout",
        _ => "Unknown",
    }
}

/// Get uptime in seconds
#[allow(dead_code)]
pub fn uptime_seconds() -> u64 {
    unsafe { esp_timer_get_time() as u64 / 1_000_000 }
}
